components/brave_wallet/hd_key.cc

```diff
@@ -0,0 +1,411 @@
+  hdkey->SetChainCode(IR);
+  return hdkey;
+}
+
+// static
+std::unique_ptr<HDKey> HDKey::GenerateFromExtendedKey(const std::string& key) {
+  std::vector<unsigned char> buf(SERIALIZATION_LEN);
+  if (!DecodeBase58Check(key, buf, buf.size())) {
+    LOG(ERROR) << __func__ << ": DecodeBase58Check failed";
+    return nullptr;
+  }
+
+  // version(4) || depth(1) || parent_fingerprint(4) || index(4) || chain(32) ||
+  // key(33)
+  uint8_t* ptr = reinterpret_cast<uint8_t*>(buf.data());
+  int32_t version = ptr[0] << 24 | ptr[1] << 16 | ptr[2] << 8 | ptr[3] << 0;
+  DCHECK(version == MAINNET_PUBLIC || version == MAINNET_PRIVATE);
+  ptr += sizeof(version);
+
+  uint8_t depth = *ptr;
+  ptr += sizeof(depth);
+
+  int32_t parent_fingerprint =
+      ptr[0] << 24 | ptr[1] << 16 | ptr[2] << 8 | ptr[3] << 0;
+  ptr += sizeof(parent_fingerprint);
+
+  int32_t index = ptr[0] << 24 | ptr[1] << 16 | ptr[2] << 8 | ptr[3] << 0;
+  ptr += sizeof(index);
+
+  std::unique_ptr<HDKey> hdkey =
+      std::make_unique<HDKey>(depth, parent_fingerprint, index);
+
+  std::vector<uint8_t> chain_code(ptr, ptr + 32);
+  ptr += chain_code.size();
+  hdkey->SetChainCode(chain_code);
+
+  if (*ptr == 0x00) {
+    DCHECK_EQ(version, MAINNET_PRIVATE);
+    std::vector<uint8_t> key(ptr + 1, ptr + 33);
+    hdkey->SetPrivateKey(key);
+  } else {
+    DCHECK_EQ(version, MAINNET_PUBLIC);
+    std::vector<uint8_t> key(ptr, ptr + 33);
+    hdkey->SetPublicKey(key);
+  }
+
+  return hdkey;
+}
+
+void HDKey::SetPrivateKey(const std::vector<uint8_t>& value) {
+  if (value.size() != 32) {
+    LOG(ERROR) << __func__ << ": pivate key must be 32 bytes";
+    return;
+  }
+  private_key_ = value;
+  GeneratePublicKey();
+  identifier_ = Hash160(public_key_);
+
+  const uint8_t* ptr = identifier_.data();
+  fingerprint_ = ptr[0] << 24 | ptr[1] << 16 | ptr[2] << 8 | ptr[3] << 0;
+}
+
+std::string HDKey::GetPrivateExtendedKey() const {
+  std::vector<uint8_t> key;
+  key.push_back(0x00);
+  key.insert(key.end(), private_key_.begin(), private_key_.end());
+  return Serialize(MAINNET_PRIVATE, key);
+}
+
+void HDKey::SetPublicKey(const std::vector<uint8_t>& value) {
+  // Compressed only
+  if (value.size() != 33) {
+    LOG(ERROR) << __func__ << ": public key must be 33 bytes";
+    return;
+  }
+  // Verify public key
+  secp256k1_pubkey pubkey;
+  if (!secp256k1_ec_pubkey_parse(secp256k1_ctx_, &pubkey, value.data(),
+                                 value.size())) {
+    LOG(ERROR) << __func__ << ": not a valid public key";
+    return;
+  }
+  public_key_ = value;
+  identifier_ = Hash160(public_key_);
+
+  const uint8_t* ptr = identifier_.data();
+  fingerprint_ = ptr[0] << 24 | ptr[1] << 16 | ptr[2] << 8 | ptr[3] << 0;
+}
+
+std::string HDKey::GetPublicExtendedKey() const {
+  return Serialize(MAINNET_PUBLIC, public_key_);
+}
+
+void HDKey::SetChainCode(const std::vector<uint8_t>& value) {
+  chain_code_ = value;
+}
+
+std::unique_ptr<HDKey> HDKey::DeriveChild(uint32_t index) {
```

Comment:
Should we validate the index is not out of bounds completely? Maybe I am missing this below but I think this could be an easy safety measure.

Reply:
Nvm. Just realized its an int 32 so that is not necessary :)